File: PageChanger/app/src/main/java/lib/core/ActivityCore.java
package lib.core;

import java.util.ArrayList;

import com.aeddang.pagechanger.R;

import lib.ViewUtil;

import android.os.Bundle;
import android.os.Handler;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.DialogInterface.OnClickListener;
import android.graphics.Typeface;
import android.util.Log;
import android.view.Gravity;
import android.view.KeyEvent;
import android.view.Menu;
import android.view.View;
import android.view.ViewGroup.LayoutParams;
import android.view.animation.Animation;
import android.view.animation.AnimationUtils;
import android.view.animation.Animation.AnimationListener;
import android.view.inputmethod.InputMethodManager;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.ProgressBar;

public class ActivityCore extends Activity implements  AnimationListener
{

	public static final int HOME=0;
	protected ViewCore currentView;
	protected FrameLayout viewer,body;
	protected ProgressBar loadingBar;
	protected ImageView dimed;
	
	
	private Animation currentAni;
	protected PageObject currentPageObj,homePageObject;
	protected ArrayList<PageObject> historyA;
    private Boolean isChangeAble,isInit;
	private static ActivityCore instence;
	
	
	private Handler removeHandler=new Handler();
	
	public static ActivityCore getInstence(){
		return instence;
	}
	public int getScreenOrientation() {
	     int rotation = getWindowManager().getDefaultDisplay().getRotation();
	     return rotation;
	}
	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		
		instence=this;
        isChangeAble=true;
        isInit=false;
        
        setContentView(R.layout.activity_main);
        loadingBar=(ProgressBar) findViewById(R.id._loadingBar);
        dimed=(ImageView) findViewById(R.id._dimed);
        body=(FrameLayout) findViewById(R.id._body);
        viewer=(FrameLayout) findViewById(R.id._viewer);
        
        Log.i("","onCreate :"+viewer);
      
        historyA=new  ArrayList<PageObject>();
     }
	
	@Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        switch (keyCode) {
                case KeyEvent.KEYCODE_BACK:
                     if(!isChangeAble){
                     	return true;
                     }
                     changeViewBack();
                break;
                 
                
         }
		 return false;
    }


    
    
    public void hideKeyBoard()
	{
		InputMethodManager imm = (InputMethodManager) this.getSystemService(Context.INPUT_METHOD_SERVICE); 
        if(imm!=null){
        	if(this.getCurrentFocus()!=null){
        		imm.hideSoftInputFromWindow(this.getCurrentFocus().getWindowToken(), 0);
        	}
        	
        }
	}
	public void changeViewMain() {
    	PageObject pobj=new PageObject(HOME);
    	pobj.dr=-1;	
    	changeView( pobj);
    }
    public void changeViewBack() {
    	
    	if(currentView!=null){
    		if(!currentView.directBack()){
        		return;
        	}
        }
    	if(currentPageObj==null)
    	{
    		changeViewBackDirect();
    	}else{
    		if(currentPageObj.isHome){
         	    viewAlertSelect("", R.string.msg_app_out, new DialogInterface.OnClickListener() {
     			public void onClick(DialogInterface dialog, int which) {
     				Log.i("witch","which="+which);
     				if(which==-1){
     					android.os.Process.killProcess(android.os.Process.myPid());
     				}else{
     					
     				}
     				
     			}
     		});
            }else{
        	   changeViewBackDirect();
            }
    	}
    	
    }
    
    public void changeViewBackDirect() {
    	
    	if(currentPageObj!=null){
    		currentPageObj.isHistory=false;
    	}
    	
    	PageObject pobj;
    	int size=historyA.size()-1;
    	if(size<0){
    		pobj=new PageObject(HOME);
    		
    	}else{
    		pobj=historyA.get(size);
    		historyA.remove(size);
    	}
    	
    	if(pobj.pageID==currentPageObj.pageID){
    		changeViewBack();
    		return;
    	}
    	pobj.dr=-1;	
    	Log.i("", "changeViewBackDirect historyA.size="+historyA.size());
    	changeView(pobj);
    }
    
    
    private void addHistory(PageObject info){
    	
    	if(!info.isHistory){
    		return;
    	}
    	if(info.isHome){
    		homePageObject=info;
    		historyA.clear();
    		return;
    	}
    	historyA.add(info);
    	Log.i("", "historyA.size="+historyA.size());
    }
    public int getCurrentPageID(){
    	
    	if(currentPageObj==null){
    		return -1;
    	}else{
    		return currentPageObj.pageID;
    		
    	}
    } 
    public void changeView(PageObject info) {
        if(!isChangeAble){
        	return;
        	
        }
        
        
        loadingStop();
        isChangeAble=false;
    	currentPageObj=info;
    	changeInit();
        creatView();
        
        if(info.isHome==true){
        	historyA.clear();
        }
        
        viewer.addView(currentView, new LayoutParams( LayoutParams.MATCH_PARENT, LayoutParams.MATCH_PARENT ));
        changeStart();  

    }
    protected void changeInit() { 
    	Animation removeAni;
        switch(currentPageObj.dr){
        case 1:
        	 currentAni = AnimationUtils.loadAnimation(this, R.anim.motion_left_in);
        	 removeAni = AnimationUtils.loadAnimation(this, R.anim.motion_right_out);
        	break;
        case -1:
       	     currentAni = AnimationUtils.loadAnimation(this, R.anim.motion_right_in);
       	     removeAni = AnimationUtils.loadAnimation(this, R.anim.motion_left_out);
       	    break;
       	default:
       		currentAni = AnimationUtils.loadAnimation(this, R.anim.motion_in);
       	    removeAni = AnimationUtils.loadAnimation(this, R.anim.motion_out);
       		break;
        	
        }
        if(currentView!=null){
        	currentView.startAnimation( removeAni );
        }
    }
    protected void changeStart() { 
    	currentAni.setAnimationListener(this);
        currentView.startAnimation( currentAni );
    }
    protected void creatView() { 
    	/*
    	switch(currentPageObj.pageID){
    	case HOME:
    		info.isHome=true;
    		currentView=new Main(this, info);
    		break;
    	}
        */
    }
    protected void changeEnd() { 
    	removeCore();
    }
    private void removeCore(){
    	
		
		int num=viewer.getChildCount();
		for(int i=0;i<num;++i){
		    ViewCore core=(ViewCore) viewer.getChildAt(i);
		    if(currentView==core){
				
			}else{
				if(core!=null){
					addHistory(core.pageInfo);
					core.removeViewCore();
					viewer.removeView(core);
				}
			}
		}
		if(currentView!=null){
			  currentView.movedInit();
		}
		isChangeAble=true;
    }
    protected void doAnimationEnd(Animation arg) { 
    	if( currentAni==arg){
			
			removeHandler.post(new Runnable() {
		        public void run() {
		        	 removeCore();
		        }
			});	
    	}
    }
    public void onAnimationEnd(Animation arg) {
		
    	doAnimationEnd(arg);
    }
    public void loadingStart(Boolean isDimed){
    	if(isDimed){
    		dimed.setVisibility(View.VISIBLE);
    		
    	}else{
    		dimed.setVisibility(View.GONE);
    	}
    	loadingBar.setVisibility(View.VISIBLE); 
    }
    public void loadingStop(){
    	 dimed.setVisibility(View.GONE);
    	 loadingBar.setVisibility(View.GONE);  
    }
    public void viewAlertSelect(String title,int msgId,OnClickListener listener) {
    	String msg=this.getString(msgId);
    	viewAlertSelect(title,msg,listener);
    	
    }
    public void viewAlert(String title,int msgId,OnClickListener listener) {
    	String msg=this.getString(msgId);
    	viewAlert(title,msg,listener);
    }
    
    public void viewAlertSelect(String title,String msg,OnClickListener listener) {
    	
    	AlertDialog.Builder  alert = new AlertDialog.Builder(this);
    	alert.setTitle(title);
    	alert.setMessage(msg);
    	alert.setPositiveButton(R.string.btn_ok, listener);
    	alert.setNegativeButton(R.string.btn_no, listener);
    	alert.show();
    }
    public void viewAlertSelectWithMenu(String title,String msg,CharSequence[] items,OnClickListener listener) {
    	
    	AlertDialog.Builder  alert = new AlertDialog.Builder(this);
    	alert.setTitle(title);
    	alert.setPositiveButton(items[0], listener);
    	alert.setNegativeButton(items[1], listener);
    	
    	alert.show();
    }
    public void viewAlert(String title,String msg,OnClickListener listener) {
    	
    	AlertDialog.Builder  alert = new AlertDialog.Builder(this);
    	alert.setTitle(title);
    	alert.setMessage(msg);
    	alert.setPositiveButton(R.string.btn_ok, listener);
        alert.show();
    }
	@Override
	public boolean onCreateOptionsMenu(Menu menu) {
		// Inflate the menu; this adds items to the action bar if it is present.
		getMenuInflater().inflate(R.menu.main, menu);
		return true;
	}
	@Override
	public void onAnimationRepeat(Animation arg0) {
		// TODO Auto-generated method stub
		
	}
	@Override
	public void onAnimationStart(Animation arg0) {
		// TODO Auto-generated method stub
		
	}

}

File: Magellan/app/src/main/java/lib/core/ActivityCore.kt
package lib.core

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.content.res.Configuration
import android.graphics.Point
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.KeyEvent
import android.view.Menu
import android.view.View
import android.view.ViewGroup
import android.view.animation.Animation
import android.view.animation.AnimationUtils
import android.view.inputmethod.InputMethodManager
import android.widget.*
import com.aeddang.magellan.R
import lib.ViewUtil
import lib.observer.ObserverController
import java.util.ArrayList

open class ActivityCore : AppCompatActivity(), Animation.AnimationListener, View.OnClickListener
{

    open var currentView: ViewCore? = null
    var body: FrameLayout? = null
    var viewer: FrameLayout? = null
    var popupViewer: FrameLayout? = null
    private var viewerDimed: Button? = null
    open var loadingBar: ProgressBar? = null
    open var loadingImage: ImageView? = null
    open var dimed: ImageView? = null

    private var currentAni: Animation? = null
    private var introAni: Animation? = null
    private var loadingAni: Animation? = null
    open var currentPageObj: PageObject? = null
    open var homePageObject: PageObject? = null
    open var historyA: ArrayList<PageObject> = ArrayList<PageObject>()
    open var popupA: ArrayList<ViewCore> = ArrayList<ViewCore>()
    open var intro: FrameLayout? = null
    open var isInit: Boolean? = false

    open var finalOrientation = -1
    open var willChangeOrientation = SCREEN_CHANGED_TYPE_NONE

    private var isChangeAble: Boolean? = null
    private var isDimed: Boolean? = null
    private val removeHandler = Handler()
    private var backKeyCount = 0


    var dpi: Float = 0.toFloat()
        get() = applicationContext.resources.displayMetrics.density

    val screenOrientation: Int
        get() = windowManager.defaultDisplay.rotation

    val appSize: Point
        get()
        {
            val rec = Point()
            if(body != null) {
                rec.x = body!!.width
                rec.y = body!!.height
            }
            return rec
        }
    val currentPageID: Int
        get() = if (currentPageObj == null)
        {
            -1
        }
        else
        {
            currentPageObj!!.pageID
        }

    override fun onCreate(savedInstanceState: Bundle?)
    {
        super.onCreate(savedInstanceState)
        if (isInit == true) return
        ActivityCore.instence = this
        isChangeAble = true
        isDimed = false
        setContentView(R.layout.activity_main)
        loadingBar = findViewById<View>(R.id._loadingBar) as ProgressBar
        //loadingImage=(ImageView) findViewById(R.id._loadingImage);
        dimed = findViewById<View>(R.id._dimed) as ImageView
        body = findViewById<View>(R.id._body) as FrameLayout
        viewer = findViewById<View>(R.id._viewer) as FrameLayout
        popupViewer = findViewById<View>(R.id._popupViewer) as FrameLayout
        intro = findViewById<View>(R.id._intro) as FrameLayout
        viewerDimed = findViewById<View>(R.id._viewerDimed) as Button
        viewerDimed?.visibility = View.GONE
        viewerDimed?.setOnClickListener(this)
        loadingAni = AnimationUtils.loadAnimation(this, R.anim.motion_rotate)
        loadingImage?.visibility = View.GONE
        actionBar?.hide()
    }

    override fun onKeyDown(keyCode: Int, event: KeyEvent): Boolean
    {
        when (keyCode) {
            KeyEvent.KEYCODE_BACK -> {
                if (isChangeAble == false || isDimed == true) return true
                //SoundManeger.getInstence().play();
                changeViewBack()
            }
        }
        return false
    }

    fun update()
    {
        super.onPause()
        currentView?.onUpdate()
        for (pop in popupA) pop.onUpdate()
    }

    override fun onPause()
    {
        super.onPause()
        currentView?.onPause()
        for (pop in popupA) pop.onPause()

    }

    override fun onResume()
    {
        super.onResume()
        currentView?.onResume()
        for (pop in popupA) pop.onResume()

    }

    override fun onClick(arg: View)
    {
        if (viewerDimed === arg) hideKeyBoard()
    }

    open fun resizeScreen(proposedWidth: Int, proposedheight: Int)
    {
        for (pop in popupA) pop.setBodySizeChanged(proposedWidth, proposedheight)

        if (willChangeOrientation >= SCREEN_CHANGED_TYPE_ORIENTATION)
        {
            if (finalOrientation != willChangeOrientation)
            {
                finalOrientation = willChangeOrientation
                ObserverController.shareInstence().notifyPost(NotificationOrientationChanged, finalOrientation, null)
                orientationChanged(finalOrientation)
            }
            willChangeOrientation = SCREEN_CHANGED_TYPE_RESIZED
        }
    }

    fun keyBoardShow(diff: Int)
    {
        viewerDimed?.visibility = View.VISIBLE
        ObserverController.shareInstence().notifyPost(NotificationKeyBoardShow, diff, null)

    }

    fun keyBoardHidden() {
        viewerDimed?.visibility = View.GONE
        ObserverController.shareInstence().notifyPost(NotificationKeyBoardHidden, null, null)

    }

    fun hideKeyBoard() {
        viewerDimed?.visibility = View.GONE
        val imm = this.getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
        if (imm != null)
        {
            if (this.currentFocus != null) imm.hideSoftInputFromWindow(this.currentFocus!!.windowToken, 0)
        }
    }

    fun changeViewMain()
    {
        val pobj = PageObject(HOME)
        pobj.dr = -1
        changeView(pobj)
    }

    fun changeViewBack()
    {
        if (popupA.size > 0) {
            val pop = popupA[popupA.size - 1]
            if (!pop.directBack()) return
            removePopup(pop)
            return
        }

        if (currentView != null)
        {
            if (currentView!!.directBack() == false) return
        }
        if (currentPageObj == null || currentPageObj!!.isHome)
        {
            backKeyCount++
            if (backKeyCount == 1)
            {
                val msgID = LanguageFactory.shareInstence().getResorceID("msg_app_out")
                Toast.makeText(this, msgID, Toast.LENGTH_LONG).show()
            }
            else
            {
                android.os.Process.killProcess(android.os.Process.myPid())
            }
        }
        else
        {
            changeViewBackDirect()
        }

    }

    fun changeViewBackDirect() {

        if(currentPageObj == null) return
        currentPageObj?.isHistory = false
        val pobj: PageObject
        val size = historyA.size - 1
        if (size < 0)
        {
            pobj = PageObject(HOME)
        }
        else
        {
            pobj = historyA[size]
            historyA.removeAt(size)
        }
        if (pobj.pageID == currentPageObj!!.pageID)
        {
            changeViewBack()
            return
        }
        pobj.dr = -1
        //Log.i("", "changeViewBackDirect historyA.size="+historyA.size());
        prevChangeViewBack(pobj)
        changeView(pobj)
    }

    open fun configurationChanged(newConfig: Configuration) {

    }

    open fun orientationChanged(orientation: Int) {

        Log.i("orientationChanged", "orientationChanged  : $orientation")
    }

    override fun onConfigurationChanged(newConfig: Configuration)
    {
        super.onConfigurationChanged(newConfig)
        if (this.screenOrientation != newConfig.orientation) willChangeOrientation = newConfig.orientation
        Log.i("willChangeOrientation", "willChangeOrientation  : $willChangeOrientation")
        if (currentView != null) currentView!!.configurationChanged(newConfig)
        configurationChanged(newConfig)
    }

    open fun prevChangeViewBack(info: PageObject) {

    }

    fun removeIntroStart(offset: Int) {


        if (intro != null)
        {
            introAni = AnimationUtils.loadAnimation(this, R.anim.motion_out)
            introAni!!.startOffset = offset.toLong()
            introAni!!.setAnimationListener(this)
            intro!!.startAnimation(introAni!!)
        }
    }

    private fun addHistory(info: PageObject) {

        if (!info.isHistory) return
        if (info.isHome)
        {
            homePageObject = info
            historyA.clear()
            return
        }
        historyA.add(info)
        //Log.i("", "historyA.size=" + historyA.size)
    }

    fun hasPopup(id: Int): Boolean
    {
        for (pop in popupA) {
            if (pop.pageInfo.pageID == id) return true
        }
        return false
    }

    fun addSinglePopup(info: PageObject): ViewCore?
    {
        return if (this.hasPopup(info.pageID) == true) null else this.addPopup(info)
    }

    fun addPopup(info: PageObject): ViewCore?
    {
        if (info.dr == 2) info.dr = 1
        val ani = getInAni(info)
        val pop = creatPopup(info) ?: return null
        backKeyCount = 0
        popupA.add(pop)
        popupViewer?.addView(pop, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT))
        pop.initCore()
        popupAniStart(pop, ani, true)
        return pop
    }

    fun removePopup()
    {
        if (popupA.size > 0) removePopup(popupA[0])
    }

    fun removePopup(pop: ViewCore)
    {
        if (popupA.indexOf(pop) != -1)
        {
            loadingStop()
            popupA.remove(pop)
            val ani = getOutAni(pop.pageInfo)
            popupAniStart(pop, ani, false)
        }

    }

    fun getInAni(pObj: PageObject): Animation
    {
        val ani: Animation
        when (pObj.dr)
        {
            1 -> ani = AnimationUtils.loadAnimation(this, R.anim.motion_up_in)
            -1 -> ani = AnimationUtils.loadAnimation(this, R.anim.motion_down_in)
            else -> ani = AnimationUtils.loadAnimation(this, R.anim.motion_in)
        }
        return ani

    }

    fun getOutAni(pObj: PageObject): Animation
    {
        val ani: Animation
        when (pObj.dr)
        {
            1 -> ani = AnimationUtils.loadAnimation(this, R.anim.motion_up_out)
            -1 -> ani = AnimationUtils.loadAnimation(this, R.anim.motion_down_out)
            else -> ani = AnimationUtils.loadAnimation(this, R.anim.motion_out)
        }
        return ani

    }

    open fun creatPopup(pinfo: PageObject): ViewCore?
    {
        /*
    	ViewCore pop;
    	switch(pinfo.pageID){
    		case Config.POPUP_SELECT_AREA:
    			pop=new PopupSelectArea(this, pinfo);
    			break;
    	}

    	return pop;
        */
        return null
    }

    open fun doPopupAnimationEnd(pop: ViewCore, isOpen: Boolean?)
    {
        if (isOpen!!) {
            pop.movedInit()
        }
        else
        {
            pop.removeViewCore()
            ViewUtil.remove(pop)
            if (popupA.size == 0) popupViewer?.visibility = View.GONE
        }
    }

    open fun popupAniStart(pop: ViewCore, ani: Animation, isOpen: Boolean)
    {
        if (isOpen == false)
        {
            pop.removeViewInit()
        }
        else
        {
            popupViewer?.visibility = View.VISIBLE
        }

        ani.setAnimationListener(object : Animation.AnimationListener
        {
            override fun onAnimationEnd(arg: Animation)
            {
                val popupHandler = Handler()
                popupHandler.post { doPopupAnimationEnd(pop, isOpen) }

            }
            override fun onAnimationRepeat(arg0: Animation) {}
            override fun onAnimationStart(arg0: Animation) {}
        })
        pop.startAnimation(ani)
    }

    /////////////////////////////////////////////////////////////////////////////////////


    fun changeView(info: PageObject)
    {
        if (isChangeAble==false) return

        backKeyCount = 0
        loadingStop()
        isChangeAble = false
        if (info.dr == 2)
        {
            if (currentPageObj != null)
            {
                if (info.pageID > currentPageObj!!.pageID)
                {
                    info.dr = 1

                }
                else if (info.pageID < currentPageObj!!.pageID)
                {
                    info.dr = -1
                }
                else
                {
                    info.dr = 0
                }
            }
            else
            {
                info.dr = 0
            }
        }
        currentPageObj = info
        changeInit()
        creatView()
        if (info.isHome == true) historyA.clear()
        viewer?.addView(currentView, ViewGroup.LayoutParams(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT))
        currentView?.initCore()
        changeStart()

    }

    open fun changeInit()
    {
        val removeAni: Animation
        when (currentPageObj!!.dr)
        {
            1 -> {
                currentAni = AnimationUtils.loadAnimation(this, R.anim.motion_left_in)
                removeAni = AnimationUtils.loadAnimation(this, R.anim.motion_right_out)
            }
            -1 -> {
                currentAni = AnimationUtils.loadAnimation(this, R.anim.motion_right_in)
                removeAni = AnimationUtils.loadAnimation(this, R.anim.motion_left_out)
            }
            else -> {
                currentAni = AnimationUtils.loadAnimation(this, R.anim.motion_in)
                removeAni = AnimationUtils.loadAnimation(this, R.anim.motion_out)
            }
        }
        currentView?.removeViewInit()
        currentView?.startAnimation(removeAni)
    }

    open fun changeStart()
    {
        currentAni?.setAnimationListener(this)
        currentView?.startAnimation(currentAni)
    }

    open fun creatView()
    {
        /*
    	switch(currentPageObj.pageID){
    	case HOME:
    		info.isHome=true;
    		currentView=new Main(this, info);
    		break;
    	}
        */
    }

    open fun changeEnd()
    {
        removeCore()
    }

    private fun removeCore()
    {
        if(viewer == null) return
        val num = viewer!!.childCount
        for (i in 0 until num) {
            val core = viewer!!.getChildAt(i) as? ViewCore
            if (currentView === core)
            {

            }
            else
            {
                if (core != null)
                {
                    addHistory(core!!.pageInfo)
                    core!!.removeViewCore()
                    viewer!!.removeView(core)
                }
            }
        }
        currentView?.movedInit()
        isChangeAble = true
    }

    private fun removeIntro()
    {
        ViewUtil.remove(intro)
        intro = null
    }

    open fun doAnimationEnd(arg: Animation)
    {
        if (currentAni === arg)
        {
            removeHandler.post { removeCore() }
        }
        else if (introAni === arg)
        {
            removeHandler.post { removeIntro() }
        }
    }


    override fun onAnimationEnd(arg: Animation) {
        doAnimationEnd(arg)
    }

    fun loadingStart(_isDimed: Boolean)
    {
        isDimed = _isDimed
        if (isDimed == true) dimed?.visibility = View.VISIBLE else dimed?.visibility = View.GONE
        loadingBar?.visibility = View.VISIBLE
        loadingImage?.animation = loadingAni
        loadingImage?.visibility = View.VISIBLE

    }

    fun loadingStop()
    {
        isDimed = false
        dimed?.visibility = View.GONE
        loadingBar?.visibility = View.GONE
        loadingImage?.animation = null
        loadingImage?.visibility = View.GONE
    }

    fun viewAlertSelect(title: String, msgId: Int, listener: DialogInterface.OnClickListener)
    {
        val msg = getString(msgId)
        viewAlertSelect(title, msg, listener)

    }

    fun viewAlert(title: String, msgId: Int, listener: DialogInterface.OnClickListener)
    {
        val msg = getString(msgId)
        viewAlert(title, msg, listener)
    }

    fun viewAlertSelect(title: String, msg: String, listener: DialogInterface.OnClickListener)
    {
        val alert = AlertDialog.Builder(this)
        alert.setTitle(title)
        alert.setMessage(msg)
        alert.setPositiveButton(R.string.btn_ok, listener)
        alert.setNegativeButton(R.string.btn_no, listener)
        alert.show()
    }

    fun viewAlertSelectWithMenu(title: String, msgId: Int, items: Array<CharSequence>, listener: DialogInterface.OnClickListener)
    {
        val msg = getString(msgId)
        viewAlertSelectWithMenu(title, msg, items, listener)
    }

    fun viewAlertSelectWithMenu(title: String, msg: String, items: Array<CharSequence>, listener: DialogInterface.OnClickListener)
    {
        val alert = AlertDialog.Builder(this)
        alert.setTitle(title)
        alert.setMessage(msg)
        alert.setPositiveButton(items[0], listener)
        alert.setNegativeButton(items[1], listener)
        alert.show()
    }

    fun viewAlert(title: String, msg: String, listener: DialogInterface.OnClickListener)
    {
        val alert = AlertDialog.Builder(this)
        alert.setTitle(title)
        alert.setMessage(msg)
        alert.setPositiveButton(R.string.btn_ok, listener)
        alert.show()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean
    {
        // Inflate the menu; this adds items to the action bar if it is present.
        //getMenuInflater().inflate(R.menu.main, menu);
        return true
    }


    override fun onAnimationRepeat(arg0: Animation)
    {
        // TODO Auto-generated method stub
    }

    override fun onAnimationStart(arg0: Animation)
    {
        // TODO Auto-generated method stub
    }

    companion object
    {
        val SCREEN_CHANGED_TYPE_ORIENTATION_RESIZED = -3
        val SCREEN_CHANGED_TYPE_ORIENTATION = 0
        val SCREEN_CHANGED_TYPE_RESIZED = -1
        val SCREEN_CHANGED_TYPE_NONE = -2
        val HOME = 0
        val PREFS_NAME="javis";
        val NotificationKeyBoardShow = "NotificationKeyBoardShow"
        val NotificationKeyBoardHidden = "NotificationKeyBoardHidden"
        val NotificationOrientationChanged = "NotificationOrientationChanged"
        private var instence: ActivityCore? = null
        fun  shareInstence(): ActivityCore?
        {
            return instence
        }

    }


}